Add lsp signature help

sudormrfbin opened the pull request:
Implements signature help, aka parameter hints:

![command](https://user-images.githubusercontent.com/23398472/156893257-698b2c1b-9172-4e45-b2da-d149be3df356.gif)

### TODO

- [x] Obstructs the completion popup
- [ ] Bar cursor in insert mode can be seen through the popup if popup overlays the cursor
- [x] Popup is not updated on backspace
- [x] `editor.lsp.signature-help` config
- [x] Keybind for manual triggering (`Ctrl-S` in insert mode)

### Future Work

1. Show all signatures for a method (useful for overloaded methods)

archseer commented on helix-term/src/commands/lsp.rs:
```diff
@@ -576,20 +579,57 @@ pub fn signature_help(cx: &mut Context) {
+            let mut popup = Popup::new(SignatureHelp::ID, contents);
+            let old_popup = compositor.find_id::<Popup<SignatureHelp>>(SignatureHelp::ID);
+            popup.set_position(old_popup.and_then(|p| p.position().copied()));
+            compositor.replace_or_push(SignatureHelp::ID, popup);
```
Interesting approach, I thought we would have to update signature_help in place and execute requests there, but this is cleaner 👍🏻

archseer replied:
How are you tracking `isRetrigger` though?

sudormrfbin replied:
Retriggers are not tracked at all right now, but all trigger characters are considered as retriggers in the spec. Trigger characters usually have `,` in them so it acts like a retrigger.

archseer reviewed (commented):
>  Obstructs the completion popup

There's a TODO item in Popup or about adding a `Bias::Above/Below` that would specify a preference for popup placement around the cursor (This is only a preference because you still want to place it below if there's no space above)

https://github.com/helix-editor/helix/blob/9bfb0caf1b4bafdac8eb964f38f7820740056fff/helix-term/src/ui/popup.rs#L63-L80.

I'd add this as a builder method similar to margin: https://github.com/helix-editor/helix/blob/9bfb0caf1b4bafdac8eb964f38f7820740056fff/helix-term/src/ui/popup.rs#L46-L49

Then you can specify the signature help to be displayed above, and completion below.


We also want to add a `editor.lsp.signature-help` config item that would allow configuring / disabling this since I can imagine some users find it distracting.

archseer commented:
Config values used by `lsp-mode.el` https://github.com/emacs-lsp/lsp-mode/blob/ddb776037fc073473e0da9d07e63f57242b212fb/lsp-mode.el#L931-L958

archseer commented on helix-term/src/commands.rs:
```diff
@@ -4010,26 +4010,15 @@ pub mod insert {
+            // lsp doesn't tell us when to close the signature help, so we request
+            // the help information again after common close triggers which should
+            // return None, which in turn closes the popup.
+            let close_triggers = &[')', ';'];
```
How does this deal with nested signature help?

```
foo(bar(|),  ... )
```

You might need to search for the matching opening brace (and track starting trigger) to determine if the closing brace matches the whole function call

sudormrfbin replied:
The popup is not _forcefully_ closed on `close_triggers`, it simply sends a normal signature help request. If the response from the server is None, we close the popup in `signature_help()`. So as long as the LSP responds correctly to the cursor position, we won't have to do any manual bookkeeping.

sudormrfbin commented:
I added `Ctrl-s` as a keybind for invoking signature help in insert mode just because, any other suggestions ?

the-mikedavis commented:
It might be easier to remember something like <kbd>Ctrl</kbd><kbd>Shift</kbd><kbd>x</kbd> but OTOH that's a lot of keystrokes. I think <kbd>Ctrl</kbd><kbd>s</kbd> is ok. <kbd>Ctrl</kbd><kbd>h</kbd> could be another good option but we'd have to kill (haha) the readline/emacs binding that's already there.

sudormrfbin commented:
`ctrl-h` would be a better keybind but it'll probably clash with backspace on a lot of terminal emulators. Switching to termwiz might be beneficial for it ability to distinguish between these too.

archseer commented:
> Switching to termwiz might be beneficial for it ability to distinguish between these too.

Even then we probably shouldn't rely on https://sw.kovidgoyal.net/kitty/keyboard-protocol/ by default since not every terminal supports it.

the-mikedavis commented on helix-view/src/editor.rs:
```diff
@@ -146,10 +146,22 @@ pub struct Config {
 }
 
-#[derive(Debug, Default, Clone, PartialEq, Serialize, Deserialize)]
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 #[serde(rename_all = "kebab-case", deny_unknown_fields)]
 pub struct LspConfig {
+    /// Display LSP progress messages below statusline
     pub display_messages: bool,
+    /// Enable automatic pop up of signature help (parameter hints)
+    pub auto_signature_help: bool,
+}
+
+impl Default for LspConfig {
+    fn default() -> Self {
+        Self {
+            display_messages: false,
+            auto_signature_help: true,
+        }
+    }
```
The defaulting here behaves strangely if you have a config with one or the other filled in, like:

```toml
[editor.lsp]
display-messages = true
```

I think because the Default here applies to the whole LspConfig. Is there a way to default the fields individually?

archseer replied:
Yeah, https://serde.rs/attr-default.html

sudormrfbin replied:
Fixed in 485d18cc.

the-mikedavis commented:
That's a bad reason to avoid a key. I think it's even a weak reason to prefer a key (e.g. `ctrl-w` that deletes a word backwords in vim in insert mode). Helix's keymap shouldn't be bound by the keys others choose.

Ctrl-Space could be interesting though as long as the terminal accepts the inputs (iirc it gives a keycode of `0x00`. idk if that gets confused with another key)

pickfire commented on book/src/keymap.md:
```diff
@@ -291,6 +291,7 @@ undo-able "save point" until you return to normal mode.
 | -----                     | -----------                 | -------                 |
 | `Escape`                  | Switch to normal mode       | `normal_mode`           |
 | `Ctrl-x`                  | Autocomplete                | `completion`            |
+| `Ctrl-s`                  | Signature help (**LSP**)    | `signature_help`        |
```
Beside what @David-Else mentioned, `ctrl-s` might be better to leave it for https://github.com/helix-editor/helix/pull/2115#issuecomment-1107958417 to keep it similar to what it does on normal mode

pickfire commented on helix-term/src/commands.rs:
```diff
@@ -2608,26 +2612,15 @@ pub mod insert {
+            // lsp doesn't tell us when to close the signature help, so we request
+            // the help information again after common close triggers which should
+            // return None, which in turn closes the popup.
+            let close_triggers = &[')', ';', '.'];
```
Would this be language specific?

sudormrfbin replied:
I don't think it's going to make much difference here since common closing characters are included here. We can always expand this list here with a few additional characters if we need to.

pickfire commented on helix-term/src/commands/lsp.rs:
```diff
@@ -565,31 +568,87 @@ pub fn goto_reference(cx: &mut Context) {
+            let language = doc
+                .language()
+                .and_then(|scope| scope.strip_prefix("source."))
+                .unwrap_or("");
```
```suggestion
                .unwrap_or_default(
                );
```

pickfire commented on helix-term/src/commands/lsp.rs:
```diff
@@ -565,31 +568,87 @@ pub fn goto_reference(cx: &mut Context) {
+                match &param.label {
+                    lsp::ParameterLabel::Simple(string) => {
+                        let start = signature.label.find(string.as_str())?;
+                        Some((start, start + string.len()))
```
I didn't look closely on this. Does this also work with double width characters like CJK? This seemed to be characters length after looking at the code (wasn't mentioned in the comment whether it is character length or terminal width length) and we don't seemed to be checking the width of `required_size`.

sudormrfbin replied:
Syntax highlighting uses char offsets so that is what we need as the final `(start, end)` value. Now, LSP sends the offsets according to the configured `OffsetEncoding` (but we have it [hardcoded to utf8](https://github.com/helix-editor/helix/blob/9a1e5d5887f1da8c2b5fa8caef77f2baff349460/helix-lsp/src/client.rs#L102)) and the same is used by the LSP for `ParameterLabel::LabelOffsets`, so we would have to convert this to char offsets:

```rust
                    lsp::ParameterLabel::LabelOffsets([start, end]) => {
                        Some((*start as usize, *end as usize))
                    }
```

`str.find()` reports the byte index of the start of the pattern, so that would have to be changed too.

Strangely, with a naive test on a bit of CJK and non ASCII text, the parameter highlighting seems to work okay:

![helix-signature-help-cjk](https://user-images.githubusercontent.com/23398472/175787514-5bac103b-f912-49e4-841d-7b85bc01c652.gif)

archseer commented:
Found one issue: We need to check for the signature help capability before sending the request. If the language server doesn't support it then "Async job failed: Unhandled method textDocument/signatureHelp" is emitted to the statusline.

the-mikedavis commented on helix-term/src/ui/editor.rs:
```diff
@@ -1192,6 +1194,13 @@ impl Component for EditorView {
                     (Mode::Insert, Mode::Normal) => {
                         // if exiting insert mode, remove completion
                         self.completion = None;
+                        // TODO: Use an on_mode_change hook to remove signature help
+                        context.jobs.callback(async {
+                            let call: job::Callback = Box::new(|_editor, compositor| {
+                                compositor.remove(SignatureHelp::ID);
+                            });
+                            Ok(call)
+                        });
```
It seems like <kbd>Escape</kbd> is closing the signature help pane but not exiting insert mode. Is that something that would be solved by doing this with a mode change hook?

andreytkachenko replied:
It would be much better to close popup and exit insert mode on `Escape`

archseer replied:
Yeah that should happen, we do the same for autocompletion

archseer commented:
#2115 will use `ctrl-s` so let's pick a different key here

the-mikedavis commented:
I found a panic when editing some Elixir code (ElixirLS has signatureHelp):

```elixir
defmodule MyMod do
  @type t :: %MyOtherMod{} # hit backspace on this closing brace in insert mode
end
```

backtrace:

```
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', helix-term/src/commands/lsp.rs:630:30
stack backtrace:
   0:     0x55d87923569c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h3e2b509ce2ce6007
   1:     0x55d878b69d8c - core::fmt::write::h753c7571fa063ecb
   2:     0x55d87922dac5 - std::io::Write::write_fmt::h2815c0519c99ba09
   3:     0x55d879238845 - std::panicking::default_hook::{{closure}}::h78d3e6cf97fc623d
   4:     0x55d8792384c3 - std::panicking::default_hook::hda898f8d3ad1a5ae
   5:     0x55d8790fa5b3 - helix_term::application::Application::run::{{closure}}::{{closure}}::h7deac8b3a1379f64
   6:     0x55d879238ef1 - std::panicking::rust_panic_with_hook::h1a5ea2d6c23051aa
   7:     0x55d879238bba - std::panicking::begin_panic_handler::{{closure}}::h07f549390938b73f
   8:     0x55d879235bc4 - std::sys_common::backtrace::__rust_end_short_backtrace::h5ec3758a92cfb00d
   9:     0x55d87923891d - rust_begin_unwind
  10:     0x55d878af2d01 - core::panicking::panic_fmt::h3a79a6a99affe1d5
  11:     0x55d878af2cc2 - core::panicking::panic_bounds_check::h449d4ff4d992b84f
  12:     0x55d878e3b666 - core::ops::function::FnOnce::call_once{{vtable.shim}}::hdf6ec64cc405e5fb
  13:     0x55d878f628e6 - helix_term::job::Jobs::handle_callback::h8f4ba4b0e15ed215
  14:     0x55d87910bc46 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hcde4836c2d207053
  15:     0x55d8790d3b0e - tokio::park::thread::CachedParkThread::block_on::hb184c4e12cdede34
  16:     0x55d8790ca679 - tokio::runtime::thread_pool::ThreadPool::block_on::h423a3da92b0b27a2
  17:     0x55d8790e983c - tokio::runtime::Runtime::block_on::h430b10508bd70571
  18:     0x55d8790d2dba - hx::main::h11474294caabc75b
  19:     0x55d8790ca4b3 - std::sys_common::backtrace::__rust_begin_short_backtrace::h6a464994d85679dc
  20:     0x55d8790e95dd - std::rt::lang_start::{{closure}}::ha330e0fbcaff3772
  21:     0x55d879234cfe - std::rt::lang_start_internal::h52e73755f77c7dd9
  22:     0x55d8790d63c2 - main
  23:     0x7f334e8cf790 - __libc_start_main
  24:     0x55d878b1fffa - _start
  25:                0x0 - <unknown>
```

pertinent section of verbose log:

```
2022-04-29T17:18:47.777 helix_lsp::transport [INFO] -> {"jsonrpc":"2.0","method":"textDocument/signatureHelp","params":{"position":{"character":38,"line":8},"textDocument":{"uri":"file:///home/michael/snip..snip/my_mod.ex"}},"id":1}
2022-04-29T17:18:47.815 helix_lsp::transport [INFO] <- {"id":1,"jsonrpc":"2.0","result":{"activeParameter":0,"activeSignature":0,"signatures":[]}}
2022-04-29T17:18:47.815 helix_lsp::transport [INFO] <- {"activeParameter":0,"activeSignature":0,"signatures":[]}
```

This kinda seems like a malformed signatureHelp response from ElixirLS, no? Seems like it should not be indicating `"activeSignature": 0` when `signatures` is empty.

the-mikedavis commented:
For the keymap of this since `C-s` is now taken: it looks like `C-X` behaves strangely - on alacritty it acts the same as `C-x` and on kitty the binding just doesn't work. It may suffer from some of the same keycode problems as `C-Backspace`. I think clobbering the existing `C-h` binding would be a decent compromise. (The existing `C-h` binding is a readline binding iirc.)

sudormrfbin commented:
> Found one issue: We need to check for the signature help capability before sending the request. If the language server doesn't support it then "Async job failed: Unhandled method textDocument/signatureHelp" is emitted to the statusline.

Fixed in 9a1e5d5.

> I noticed testing this that Backspace in insert mode will pop up the "Language server not active for current buffer" status message for buffers like scratch or text. Probably not a big deal but it does grab the eye a bit.

Fixed in f1de34c.

> This kinda seems like a malformed signatureHelp response from ElixirLS, no? Seems like it should not be indicating `"activeSignature": 0` when `signatures` is empty.

That is indeed an invalid response from the server, but I've noticed that `clangd` does the same thing and probably so does many others. 0a4b5b2 changes the code to be more defensive against the response.

> I think clobbering the existing `C-h` binding would be a decent compromise. (The existing `C-h` binding is a readline binding iirc.)

I'm not sure if `C-h` is interpreted as backspace (in the same way `C-o` is treated as Tab), if so we would have to find something else. It works pretty good in kitty though.

Houkime commented:
Played with it. Looks mostly alright, with the major pain point being the need to press esc 2 times to return to normal.
And if one hits any letter (like a command) after hitting esc, the autocompletion  re-popsup and, now you need to both clean up the mess and hit esc 2 times.
A lesser pain point is that with larger font (have weak eyes) it is often that full doc does not fit above, and it starts to try fit below. 
My suggestion would be to keep it simple and to just show function signature.

archseer commented on helix-lsp/src/client.rs:
```diff
@@ -645,7 +655,14 @@ impl Client {
+    ) -> Option<impl Future<Output = Result<Value>>> {
+        let capabilities = self.capabilities.get().unwrap();
+
+        #[allow(clippy::question_mark)]
+        if capabilities.signature_help_provider.is_none() {
+            return None;
+        }
```
use the question mark and add a comment right before? "return early if signature help isn't supported"

archseer commented on helix-term/src/compositor.rs:
```diff
@@ -132,6 +132,14 @@ impl Compositor {
     }
 
+    pub fn remove(&mut self, id: &'static str) -> Option<Box<dyn Component>> {
+        let idx = self
+            .layers
+            .iter()
+            .position(|layer| layer.id() == Some(id))?;
```
let idx = self.find_id(id);

sudormrfbin replied:
`find_id(id)` returns the actual component that has the specified id. A better name for the function would have been `find_by_id()`.

sudormrfbin commented on helix-term/src/ui/popup.rs:
```diff
@@ -11,12 +11,19 @@ use helix_view::graphics::{Margin, Rect};
 // a width/height hint. maybe Popup(Box<Component>)
 
+#[derive(PartialEq)]
+pub enum PositionBias {
+    Above,
+    Below,
+}
```
Reused the [`Open` enum](https://github.com/sudormrfbin/helix/blob/9a1e5d5887f1da8c2b5fa8caef77f2baff349460/helix-term/src/commands.rs#L2347-L2350) which was already being used opening lines above or below with `o` or `O`.

sudormrfbin commented:
> A lesser pain point is that with larger font (have weak eyes) it is often that full doc does not fit above, and it starts to try fit below. My suggestion would be to keep it simple and to just show function signature. Full doc also obstructs code when i would like to see the names of variables i defined above.

@Houkime I'll add a config option to control the display of docs 👍🏾

sudormrfbin commented:
Escape now closes the popup _and_ exits insert mode, similar to the completion popup. Closing the popup only can be done with `ctrl-c`. Also added a config option `editor.lsp.display-signature-help-docs` that controls the docs display.

lazytanuki commented:
Hi and thanks for this feature !

Been playing with it a little bit, found something that looks like an issue: when writing a closure and hitting `Backspace`, the signature help shows up for the function for which the closure is an argument (may not be clear, see screenshot below). You should be able to reproduce it by editing `helix-term/src/commands/lsp.rs` as shown below:

![image](https://user-images.githubusercontent.com/43273245/176023997-308d5203-3051-440c-af2a-7efac8cc9f5f.png)


Also, a keybinding suggestion: how about mapping `C-u` and `C-d` (which are already quite present in different parts of the editor) to put the signature help popup above or below current line (when it's already being displayed, not to trigger it) ? User case is when editing code in the middle of a function, signature help pops up below the current line, but you want to see what's there so you just press `C-u` and it makes the signature help switch to being above the current line.

the-mikedavis commented:
Backspace triggers the LSP call - showing the signature for the function in that case is up to rust-analyzer. It can be a bit odd to show the signature of the outer call when you're in a closure - maybe we should open an issue in rust-analyzer?

lazytanuki commented:
I am not sure about that, it is not doing it in neovim with the [lsp signature plugin](https://github.com/ray-x/lsp_signature.nvim/). When specifically asking for the signature help when inside the closure, this plugin returns "No signature help available". I'll try to see how this is handled by that plugin. Maybe a treesitter scope check ?

sudormrfbin commented:
> Also, a keybinding suggestion: how about mapping `C-u` and `C-d` (which are already quite present in different parts of the editor) to put the signature help popup above or below current line (when it's already being displayed, not to trigger it)

`C-u` and `C-d` are already taken for scrolling the popup, and I think having a single keybind that toggles top and bottom orientations would be better anyway. And it would fit better as a generic keybind added to any popup rather than only the signature help, so a separate PR will be better (after this one gets merged, since position biasing is introduced here).

pickfire commented on helix-view/src/editor.rs:
```diff
@@ -162,10 +162,25 @@ pub struct Config {
 pub struct LspConfig {
+    /// Display LSP progress messages below statusline
     pub display_messages: bool,
+    /// Enable automatic pop up of signature help (parameter hints)
+    pub auto_signature_help: bool,
```
Is there a way to manually trigger signature help? We can manually trigger completion with `ctrl-x`, maybe once we trigger that it automatically triggers signature help as well?

sudormrfbin replied:
I added a keybind earlier but [removed it](https://github.com/helix-editor/helix/pull/1755#issuecomment-1111196004) due to a conflict. Adding a keybind is as easy as binding it to `signature_help` command, and we definitely need a built-in insert mode keybind too. Any suggestions?

> We can manually trigger completion with `ctrl-x`, maybe once we trigger that it automatically triggers signature help as well?

I'd rather have a separate key that can toggle just the signature help popup since that allows exclusive control.

pickfire commented on helix-term/src/commands/lsp.rs:
```diff
@@ -627,31 +629,116 @@ pub fn goto_reference(cx: &mut Context) {
+        move |editor, compositor, response: Option<lsp::SignatureHelp>| {
+            let config = &editor.config();
+
+            if !(config.lsp.auto_signature_help
+                || SignatureHelp::visible_popup(compositor).is_some()
+                || was_manually_invoked)
+            {
```
When signature help shows up, I pressed escape and insert, signature help no longer shows up.

sudormrfbin replied:
Fixed in b675243!

sudormrfbin commented:
> I was trying this branch out and noticed that the `editor.auto-completion = false` no longer works as expected, even when I did not manually trigger the auto completion it still shows up.

I can't reproduce this; I tried both `:set auto-completion false` inside the editor and `editor.auto-completion = false` in the TOML config and neither of them show the completion popup unless manually triggered. The signature help pops up regardless of the auto completion setting, which is as expected.
